internal/endpoint/endpoint.go

```diff
@@ -2,59 +2,80 @@
 //
 // Copyright (C) 2018 IOTech Ltd
 // Copyright (c) 2019 Intel Corporation
+// Copyright (c) 2020 Dell Inc.
 //
 // SPDX-License-Identifier: Apache-2.0
 
 package endpoint
 
 import (
+	"context"
 	"fmt"
-	"os"
 	"sync"
 	"time"
 
-	"github.com/edgexfoundry/go-mod-core-contracts/clients/types"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/interfaces"
 	"github.com/edgexfoundry/go-mod-registry/registry"
 )
 
 type Endpoint struct {
-	RegistryClient registry.Client
-	passFirstRun   bool
-	WG             *sync.WaitGroup
+	ctx            context.Context
+	wg             *sync.WaitGroup
+	registryClient registry.Client
+	serviceKey     string // The key of the service as found in the service registry (e.g. Consul)
+	path           string // The path to the service's endpoint following port number in the URL
+	interval       int    // The interval in milliseconds governing how often the client polls to keep the endpoint current
+
+}
+
+func New(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	registryClient registry.Client,
+	serviceKey string,
+	path string,
+	interval int) *Endpoint {
+
+	return &Endpoint{
+		ctx:            ctx,
+		wg:             wg,
+		registryClient: registryClient,
+		serviceKey:     serviceKey,
+		path:           path,
+		interval:       interval,
+	}
 }
 
-func (e Endpoint) Monitor(params types.EndpointParams) chan string {
-	ch := make(chan string, 1)
-	ch <- e.fetch(params)
+func (e Endpoint) Monitor() chan interfaces.URLStream {
+	ch := make(chan interfaces.URLStream)
+	ticker := time.NewTicker(time.Millisecond * time.Duration(e.interval))
+	e.wg.Add(1)
 	go func() {
+		defer e.wg.Done()
+
+		// run fetchURL once before looping so we get the first check before the first timer interval
+		e.fetchURL(ch)
```

Comment:
Maybe it's better to call `e.wg.Done()` after first check ?

Reply:
Using defer instead of calling it explicitly allows it to run even when an error occurs. If we called it under the ctx.Done() case, it would only finish on the happy path. This way, it finishes whenever the routine ends, whether that's because it was successful or because it failed.

Reply:
my idea is related to the above comment: if we keep `waitGroup.wait()` we should call `wg.Done()` immediately after first check so that the main thread won't be blocked and then we can still monitor the URL change in the background

Reply:
What should we do in the error case? Never call `wg.Done()`?